oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtEncoder.java

```diff
@@ -369,4 +388,295 @@ private static URI convertAsURI(String header, URL url) {
+		 */
+		public KeyPairJwtEncoderBuilder signatureAlgorithm(SignatureAlgorithm signatureAlgorithm) {
+			Assert.notNull(signatureAlgorithm, "signatureAlgorithm cannot be null");
+			this.jwsAlgorithm = JWSAlgorithm.parse(signatureAlgorithm.getName());
+			return this;
+		}
+
+		/**
+		 * Sets the key ID ({@code kid}) to be included in the JWK and potentially the JWS
+		 * header.
+		 *
+		 * @param keyId the key identifier
+		 * @return this builder instance for method chaining
+		 */
+		public KeyPairJwtEncoderBuilder keyId(String keyId) {
+			this.keyId = keyId;
+			return this;
+		}
+
+		/**
+		 * Configures the {@link Converter} used to select the JWK when multiple keys match
+		 * the header criteria. This is generally not needed for single-key setups but is
+		 * provided for consistency.
+		 *
+		 * @param jwkSelector the {@link Converter} to select a {@link JWK}
+		 * @return this builder instance for method chaining
+		 * @since // TODO: Update version
+		 */
+		public KeyPairJwtEncoderBuilder jwkSelector(Converter<List<JWK>, JWK> jwkSelector) {
+			Assert.notNull(jwkSelector, "jwkSelector cannot be null");
+			this.jwkSelector = jwkSelector;
+			return this;
+		}
+
+		/**
+		 * Builds the {@link NimbusJwtEncoder} instance.
+		 *
+		 * @return the configured {@link NimbusJwtEncoder}
+		 * @throws IllegalStateException if the key type is unsupported or the configured
+		 *                               JWS algorithm is not compatible with the key type.
+		 * @throws JwtEncodingException  if the key is invalid (e.g., EC key with unknown curve)
+		 */
+		public NimbusJwtEncoder build() {
+			this.keyId = this.keyId != null ? this.keyId : UUID.randomUUID().toString();
+			JWK jwk = buildJwk();
+			JWKSource<SecurityContext> jwkSource = new ImmutableJWKSet<>(new JWKSet(jwk));
+			NimbusJwtEncoder encoder = this.jwsHeader  == null ?
+					new NimbusJwtEncoder(jwkSource) :
+					new NimbusJwtEncoder(jwkSource, this.jwsHeader);
+			if (this.jwkSelector != null) {
+				encoder.setJwkSelector(this.jwkSelector);
+			} else {
+				encoder.setJwkSelector(jwkSet -> jwkSet.stream().findFirst().orElse(null));
+			}
+			return encoder;
+		}
+
+		private JWK buildJwk() {
+			if (this.keyPair.getPrivate() instanceof RSAPrivateKey) {
+				RSAKey rsaKey = buildRsaJwk();
+				this.jwsHeader = JwsHeader.with(SignatureAlgorithm.from(jwsAlgorithm.getName())) // Use Spring's enum
+						.keyId(rsaKey.getKeyID())
+						.build();
+				return rsaKey;
+			} else if (this.keyPair.getPrivate() instanceof ECPrivateKey) {
+				ECKey ecKey = buildEcJwk();
+				this.jwsHeader = JwsHeader.with(SignatureAlgorithm.from(jwsAlgorithm.getName())) // Use Spring's enum
+						.keyId(ecKey.getKeyID())
+						.build();
+				return ecKey;
+			} else {
+				throw new IllegalStateException("Unsupported key pair type: " + this.keyPair.getPrivate().getClass().getName());
+			}
+		}
+
+		private RSAKey buildRsaJwk() {
+			if (this.jwsAlgorithm == null) {
```

Comment:
Note that if the signature algorithm is verified in the `signatureAlgorithm` method, then this check and the `Assert#state` check are unneeded here.